feat(mobile): add bulk download functionality

JobiJoba opened the pull request:
- Added `downloadAll` method to `IDownloadRepository` and its implementation in `DownloadRepository` to handle multiple asset downloads.
- Implemented `downloadAllAsset` in `DownloadStateNotifier` to trigger bulk downloads.
- Updated `DownloadService` to create download tasks for all selected assets.
- Enhanced UI with new download success and failure messages in `en.json`.
- Added download button to `ControlBottomAppBar` and integrated download functionality in `MultiselectGrid`.

## Description

Feature request from #5068 

Few decision has been taken: 

1. Allow to download a file only if he's present remotely (even if he's already present locally also) ; This is a feature that I often use when I lost a picture but could find it in Google Photos, by downloading it; it create a copy but at the top of the stack of my photos. 
2. **In the gallery asset; we can only download a picture if it's remote and non-local; I would like to change this to be the same as the multi-download**
3. Because we can select multiple state of an asset, sometimes it can be remote only, local only, remote and local; I've adapted the message to show a "download (count) files successful - (error) files failed" 

I've made a small typo fix which is not really related to this issue here:

```dart
      'SelectionAssetState(hasRemote: $hasRemote, hasMerged: $hasMerged, hasMerged: $hasMerged, selectedCount: $selectedCount)';

   'SelectionAssetState(hasRemote: $hasRemote, hasLocal: $hasLocal, hasMerged: $hasMerged, selectedCount: $selectedCount)';
``` 
(hasMerged was there two time instead of hasLocal) 


-- 

I think how it works currently is ok but in a better world I would like to also let people download a picture that is only local. 
What do you think? 

Fixes # (issue)

## How Has This Been Tested?

- Select multiple or one asset(s) which are present remotely
- Check that the download button is present
- Click on the download button
- Check if the images are in your local library

-- 

- Select a non remote asset
- Check that the download button is not present

-- 

- Select a non remote and a remote asset
- Check that the button download is present
- Check that the remote image is well downloaded.

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

<!-- Images go below this line. -->
![92083](https://github.com/user-attachments/assets/0fa12849-a37f-4042-aa76-048a90fbf01b)

</details>

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

JobiJoba committed:
feat(mobile): add bulk download functionality and update UI messages

- Added `downloadAll` method to `IDownloadRepository` and its implementation in `DownloadRepository` to handle multiple asset downloads.
- Implemented `downloadAllAsset` in `DownloadStateNotifier` to trigger bulk downloads.
- Updated `DownloadService` to create download tasks for all selected assets.
- Enhanced UI with new download success and failure messages in `en.json`.
- Added download button to `ControlBottomAppBar` and integrated download functionality in `MultiselectGrid`.

alextran1502 commented on i18n/en.json:
```diff
@@ -461,6 +461,8 @@
   "assets_added_to_album_count": "Added {count, plural, one {# asset} other {# assets}} to the album",
   "assets_added_to_name_count": "Added {count, plural, one {# asset} other {# assets}} to {hasName, select, true {<b>{name}</b>} other {new album}}",
   "assets_count": "{count, plural, one {# asset} other {# assets}}",
+  "assets_downloaded_successfully": "Downloaded {count} file(s) successfully",
```
You can use the ICU message format for singular and plural similar to the one above, `assets_count`

JobiJoba replied:
I tried to make that syntaxe worked but couldn't -_- 

I didn't find any example of this syntaxe used in dart but it's used in Svelte; the way it is done in Easy_localization is a bit different and I'm afraid it will broke the translation system if we use that

Ex from the doc:
```json
{
  "day": {
    "zero":"{} дней",
    "one": "{} день",
    "two": "{} дня",
    "few": "{} дня",
    "many": "{} дней",
    "other": "{} дней"
  },
  "money": {
    "zero": "You not have money",
    "one": "You have {} dollar",
    "many": "You have {} dollars",
    "other": "You have {} dollars"
  },
  "money_args": {
    "zero": "{} has no money",
    "one": "{} has {} dollar",
    "many": "{} has {} dollars",
    "other": "{} has {} dollars"
  },
  "money_named_args": {
    "zero": "{name} has no money",
    "one": "{name} has {money} dollar",
    "many": "{name} has {money} dollars",
    "other": "{name} has {money} dollars"
  }
}
```

JobiJoba replied:
done, thanks for the tips :)

alextran1502 reviewed (approved):
Very clear and concise code! Thanks

shenlong-tanwen commented on mobile/lib/services/download.service.dart:
```diff
@@ -159,9 +159,22 @@ class DownloadService {
+
   Future<void> download(Asset asset) async {
+    final tasks = _createDownloadTasks(asset);
+    await _downloadRepository.download(tasks.first);
```
This is wrong. You are returning a list of tasks from the `_createDownloadTasks` method and this only downloads the first task from it. For example, on iOS, if an asset is a live photo, the method returns two tasks but this would only download the static part of it. The previous implementation makes two calls to the `download` method of the repository.

This should actually be calling `_downloadRepository.downloadAll` as well

shenlong-tanwen commented on mobile/lib/services/download.service.dart:
```diff
@@ -159,9 +159,22 @@ class DownloadService {
   }
 
+  Future<List<bool>> downloadAll(List<Asset> assets) async {
+    List<DownloadTask> tasks = [];
+    for (var asset in assets) {
+      tasks.addAll(_createDownloadTasks(asset));
+    }
+    return await _downloadRepository.downloadAll(tasks);
```
```suggestion
    return await _downloadRepository
        .downloadAll(assets.expand(_createDownloadTasks).toList());
```

JobiJoba replied:
Nice shorten way; thanks

shenlong-tanwen commented on mobile/lib/services/download.service.dart:
```diff
@@ -185,16 +195,20 @@ class DownloadService {
+      ];
+    }
+
+    if (asset.remoteId == null || asset.remoteId! == "N/A") {
```
I don't think we ever assign "N/A" as the remoteId so it isn't necessary to have the second condition

JobiJoba replied:
I've seen it in a debug log that's why I did it. Let me check if I can see it again

JobiJoba replied:
ok got it :) I'll remove it then

shenlong-tanwen reviewed (changes requested):
Kindly address the comments and update the translation strings to use ICU message format. Thanks for the contribution

JobiJoba committed:
translations use i18n method t()

JobiJoba committed:
Update mobile/lib/services/download.service.dart

Co-authored-by: shenlong <139912620+shenlong-tanwen@users.noreply.github.com>

JobiJoba committed:
Merge branch 'download-in-multi-select-files' of https://github.com/JobiJoba/immich into download-in-multi-select-files

JobiJoba committed:
fix(mobile): update download logic in DownloadService

- Changed the download method to utilize downloadAll for handling multiple tasks.
- Simplified remoteId check by removing unnecessary condition.

JobiJoba commented:
All change from @shenlong-tanwen  done; thanks for the review

JobiJoba committed:
Merge branch 'main' into download-in-multi-select-files

JobiJoba committed:
Merge branch 'main' into download-in-multi-select-files

JobiJoba commented:
I'm not sure what makes the i18 fail - someone does have an idea?

shenlong-tanwen commented:
> I'm not sure what makes the i18 fail - someone does have an idea?

The keys in the translation file has to be sorted

JobiJoba committed:
sort i18n keys

JobiJoba committed:
remove the download signature from interface and logic as we use the downloadAll now

JobiJoba commented:
I was thinking the same, want me to look at how to add the progress bar? (I've tried to add the component but it didn't work but I didn't really look into it)

akostadinov commented:
Thank you for the great feature. A couple of suggestions.

I think that downloading a copy by default is suboptimal. Why not make it a non-default user preference or have a separate `Download a copy` button?

Also showing the file at the top of the stack, does it mean mtime is being messed up?
